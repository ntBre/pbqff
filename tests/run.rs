use std::{fs::read_to_string, path::Path};

use assert_cmd::Command;
use insta::{assert_snapshot, with_settings};
use tempfile::tempdir;
use test_case::test_case;

#[test_case("testfiles/test.toml", &["testfiles/intder.in"]; "sic")]
#[test_case("testfiles/cart.toml", &[]; "cart")]
#[test_case("testfiles/norm.toml", &[]; "norm")]
fn run(path: &str, other_files: &[&str]) -> std::io::Result<()> {
    let config_file = Path::new(path);
    let dir = tempdir()?;
    std::fs::copy(config_file, dir.path().join("pbqff.toml"))?;
    for file in other_files {
        let p = Path::new(file);
        let filename = p.file_name().unwrap();
        std::fs::copy(file, dir.path().join(filename))?;
    }
    let mut cmd = Command::cargo_bin("pbqff").unwrap();
    let assert = cmd.arg("pbqff.toml").current_dir(&dir).assert();
    let output = assert.get_output();

    assert!(
        output.status.success(),
        "stderr: {}\nlog: {}",
        String::from_utf8_lossy(&output.stderr),
        read_to_string(dir.path().join("pbqff.log"))?,
    );

    // filter out essentially all of the numerical results just to check that
    // all of the sections are there
    with_settings!({
        filters => vec![
            (r"(?m)^PID: \d+$", "PID: [PID]"),
            (r"(?m)^version: [a-z0-9]+$", "version: [version]"),
            (r"(?s)(normalized geometry:).*(point group)", "$1[Normalized Geometry]\n$2"),
            (r"(?ms)^Normal Coordinates:.*?^$", "[Normal Coordinates]"),
            (r"(?ms)^STATE NO.*?^$", "[Vibrational States]"),
            (r"(?s)(Transformed Geometry \(Ã…\):).*(Equilibrium)", "$1\n[Transformed Geometry]\n$2"),
            (r"(?ms)(Equilibrium Rotational Constants \(cm-1\):).*?^$", "$1\n[Rotational Constants]\n"),
            (r"(?s)(Geometry:).*(Vibrational)", "$1\n[Geometry]\n$2"),
            (r"(?s)(Vibrational Frequencies \(cm-1\):).*?(ZPT)", "$1\n[Freqs]\n$2"),
            (r"(?ms)(^Rotational Constants \(cm-1\):).*?(Coriolis Resonances)", "$1\n[Rotational Constants]\n$2"),
            (r"(?m)^(anpass sum of squared residuals):(.+)$", "$1: [residual]"),
        ],
        snapshot_suffix => path

    }, {
        assert_snapshot!(read_to_string(dir.path().join("pbqff.out")).unwrap());
    });

    Ok(())
}
