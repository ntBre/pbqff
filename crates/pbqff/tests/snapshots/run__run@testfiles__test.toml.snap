---
source: tests/run.rs
expression: "read_to_string(dir.path().join(\"pbqff.out\")).unwrap()"
snapshot_kind: text
---
PID: [PID]
version: [version]

Configuration Options:
geometry = {
C
C 1 CC
C 1 CC 2 CCC
H 2 CH 1 HCC 3 180.0
H 3 CH 1 HCC 2 180.0

CC =                  1.42101898
CCC =                55.60133141
CH =                  1.07692776
HCC =               147.81488230

}
optimize = true
charge = 0
step_size = 0.005
coord_type = sic
template = scfcrt=1.D-21 aux(precision=14 comp xp xs xw) PM6 THREADS=1
hybrid_template = scfcrt=1.D-21 aux(precision=14 comp xp xs xw) PM6 THREADS=1
queue_template = 
program = mopac
queue = local
sleep_int = 2
job_limit = 2048
chunk_size = 1
findiff = false
check_int = 100
dummy_atoms = None
norm_resume_hff = false

Optimized Geometry:
[Optimized Geometry]

Normalized Geometry:[Normalized Geometry]
Point Group = C2v(Axis(Y), Plane(Axis(Y), Axis(Z)), Plane(Axis(X), Axis(Y)))

Symmetry Internal Coordinates:
S0 (A1) = r(2-3)
S1 (A1) = r(1-2) + r(1-3)
S2 (A1) = r(2-4) + r(3-5)
S3 (A1) = ∠(4-2-1) + ∠(5-3-1)
S4 (B2) = r(1-2) - r(1-3)
S5 (B2) = r(2-4) - r(3-5)
S6 (B2) = ∠(4-2-1) - ∠(5-3-1)
S7 (B1) = τ(4-2-1-3) - τ(5-3-1-2)
S8 (A2) = τ(4-2-1-3) + τ(5-3-1-2)

5 atoms require 1585 jobs
anpass sum of squared residuals: [residual]

Refit Geometry
[Refit Geometry]
[Vibrational States]
Transformed Geometry (Å):
[Transformed Geometry]
Equilibrium Rotational Constants (cm-1):
[Rotational Constants]

Geometry:
[Geometry]
Vibrational Frequencies (cm-1):
[Freqs]
ZPT =   6673.9

Rotational Constants (cm-1):
[Rotational Constants]
Coriolis Resonances:
    4    3    C
    5    4    A
    6    5    A
    7    6    B
    8    5    B
    8    7    A

Type 1 Fermi Resonances:
    5    2
    6    2
    7    2
    8    2

Type 2 Fermi Resonances:
    3    2    0
    4    2    1
    6    2    1
    6    4    2
    8    2    0

normal termination of pbqff
